File: app/src/main/java/com/pr0gramm/app/util/SimpleJobCreator.java
package com.pr0gramm.app.util;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobCreator;
import com.google.common.base.Supplier;

public abstract class SimpleJobCreator implements JobCreator {
    private final String tag;

    protected SimpleJobCreator(String tag) {
        this.tag = tag;
    }

    @Override
    public final Job create(String s) {
        if (tag.equals(s)) {
            return create();
        } else {
            return null;
        }
    }

    /**
     * Create the real job.
     */
    protected abstract Job create();

    public static JobCreator forSupplier(String tag, Supplier<? extends Job> creator) {
        return new SimpleJobCreator(tag) {
            @Override
            protected Job create() {
                return creator.get();
            }
        };
    }
}

File: app/src/main/java/com/pr0gramm/app/util/SimpleJobCreator.kt
package com.pr0gramm.app.util

import com.evernote.android.job.Job
import com.evernote.android.job.JobCreator

abstract class SimpleJobCreator protected constructor(private val tag: String) : JobCreator {
    override fun create(s: String): Job? {
        if (tag == s) {
            return create()
        } else {
            return null
        }
    }

    /**
     * Create the real job.
     */
    protected abstract fun create(): Job

    companion object {
        inline fun forSupplier(tag: String, crossinline creator: () -> Job): JobCreator {
            return object : SimpleJobCreator(tag) {
                override fun create(): Job {
                    return creator()
                }
            }
        }
    }
}
